core/ppu_render.go

```diff
@@ -112,3 +178,58 @@ func (ppu *PPU) Sprites() []image.Image {
 	}
 	return images
 }
+
+func (ppu *PPU) bgToImage(bgIndex uint8) image.Image {
+	bg := ppu.backgroundData.bg[bgIndex]
+
+	//create an image to fit the background
+	sizeInTile := uint16(64)
+	hTileSize, vTileSize := bg.tileSize()
+	hSize := sizeInTile * hTileSize
+	vSize := sizeInTile * vTileSize
+	img := image.NewRGBA(image.Rectangle{
+		Min: image.Point{},
+		Max: image.Point{X: int(hSize), Y: int(vSize)},
+	})
+
+	//fill the image
+
+	//go through the background tiles
+	for yBgTile := uint16(0); yBgTile < uint16(sizeInTile); yBgTile++ {
+		for xBgTile := uint16(0); xBgTile < uint16(sizeInTile); xBgTile++ {
+
+			//get the background tile at these coordinates
+			bgTile := ppu.tileFromBackground(bgIndex, xBgTile, yBgTile)
+
+			// Loop over all the base tiles contained in the background tile
+			for yTile := uint16(0); yTile < bgTile.vSize/TILE_SIZE; yTile++ {
+				for xTile := uint16(0); xTile < bgTile.hSize/TILE_SIZE; xTile++ {
+
+					tile := bgTile.tileAt(xTile, yTile)
+
+					// Loop over all the pixels in the current tile
+					for y := uint16(0); y < bgTile.vSize; y++ {
+						for x, color := range ppu.tileRowColor(tile, y) {
+							if !color.Transparent {
+								img.Set(int(xBgTile*bgTile.hSize+xTile*TILE_SIZE+uint16(x)), int(yBgTile*bgTile.vSize+yTile*TILE_SIZE+uint16(y)), color)
```

Comment:
that's a really long line : 😅

Reply:
Ahah true I got myself carried the maths